### PR TITLE
[Promotion] Implement custom promotion action factory

### PR DESCRIPTION
**Pull request by Zales0123**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets |
| License       | MIT
| Doc PR        |

According to discussion about [promotion scenarios](https://github.com/Sylius/Sylius/pull/4199#issuecomment-185731781), there should be a custom promotion action factory with specialized methods.

### CODE REVIEW

File: src/Sylius/Component/Promotion/Factory/ActionFactory.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+{
+    /**
+     * @var FactoryInterface
+     */
+    private $actionFactory;
+
+    /**
+     * @param FactoryInterface $actionFactory
+     */
+    public function __construct($actionFactory)
```
Comment (pamil) [inline src/Sylius/Component/Promotion/Factory/ActionFactory.php]: Missing typehint

File: src/Sylius/Component/Promotion/Factory/ActionFactoryInterface.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+interface ActionFactoryInterface extends FactoryInterface
+{
+    /**
+     * @param int $amount
+     *
+     * @return PromotionInterface
```
Comment (pamil) [inline src/Sylius/Component/Promotion/Factory/ActionFactoryInterface.php]: Not an action?

File: src/Sylius/Component/Promotion/Factory/ActionFactory.php (hunk 1 above)
Comment (Zales0123) [inline src/Sylius/Component/Promotion/Factory/ActionFactory.php]: ozofnmp

File: src/Sylius/Component/Promotion/Factory/ActionFactoryInterface.php (hunk 2 above)
Comment (Zales0123) [inline src/Sylius/Component/Promotion/Factory/ActionFactoryInterface.php]: Obviously an action!

File: src/Sylius/Bundle/PromotionBundle/DependencyInjection/SyliusPromotionExtension.php [hunk 3]
```diff
@@ -55,14 +72,18 @@ public function load(array $config, ContainerBuilder $container)
...
+    /**
+     * @param ContainerBuilder $container
+     */
+    private function overwriteActionFactory(ContainerBuilder $container)
+    {
+        $oldActionFactory = $container->getDefinition('sylius.factory.promotion_action');
+        $newActionFactoryDefinition = new Definition(ActionFactory::class);
+
+        $actionFactory = $container->setDefinition('sylius.factory.promotion_action', $newActionFactoryDefinition);
+        $actionFactory->addArgument($oldActionFactory);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/PromotionBundle/DependencyInjection/SyliusPromotionExtension.php]: The `ActionFactory::class` should be configured in the `Configuration` class and the overwrite should look like this:
```
$baseFactoryDefinition = new Definition(Factory::class, [new Parameter('sylius.model.promotion_action.class')]);
$promotionActionFactoryClass = $container->getParameter('sylius.factory.promotion_action.class');
$decoratedPromotionActionFactoryDefinition = new Definition($promotionActionFactoryClass, [$baseFactoryDefinition]);

$container->setDefinition('sylius.factory.promotion_action', $decoratedPromotionActionFactoryDefinition);
```

File: src/Sylius/Component/Promotion/Factory/ActionFactory.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class ActionFactory implements ActionFactoryInterface
+{
+    /**
+     * @var FactoryInterface
+     */
+    private $actionFactory;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Promotion/Factory/ActionFactory.php]: `decoratedFactory`

**Comment by pjedrzejewski** [comment]
Thanks Mateusz!

**Pull request merged**